flaml/automl/model.py

```diff
@@ -1588,8 +1581,6 @@ def _callback(self, start_time, deadline, free_mem_ratio, env) -> None:
         now = time.time()
         if env.iteration == 0:
             self._time_per_iter = now - start_time
-        if now + self._time_per_iter > deadline:
-            raise EarlyStopException(env.iteration, env.evaluation_result_list)

```

Comment:
This helps control the time budget. In the case of unreproducibility, did you observe that the post-training loss is lower than the training loss?

Reply:
Hi @thinkall, I've tried investigating this issue further, and I'm a little confused. Apologies for the wall of text.

Firstly, in response to this comment, I re-added the if statement in question back to my local code, and reran `test_reproducibility_of_lgbm_regression_model` . My first result was as follows:

![image](https://github.com/user-attachments/assets/a4828041-c2f4-4abc-83a4-7495aedbe763)

The reproduced loss is lower than the original loss. In this case, we'd be best off letting the LGBM run to completion.

I then started looking into this early stopping a little more. I tracked when the callback was called, and also when the above early stopping was called. Re-running `test_reproducibility_of_lgbm_regression_model` revealed the following:

- 10 LGBM estimators were tried and tested in total
- The early stopping was called 9 times in total, once for each fold on the very last model (that makes sense).
- **The callback called was 36 times, per model - why is it being called 4 times per fold?**
- **The deadline changes each time a new model is tried - shouldn't there be a single deadline for all models, dictated by the time budget?** I imagine I'm missing something obvious on these last two points.

In summary, I see the trade-offs as being along these lines:

**In favour of Early Stopping**

1. Time budget is respected more.
2. With early stopping, we _maybe_ provide a minimum loss for that model on that fold, and we can potentially surpass that loss if we retrain the model on the same folds, with no time budget/risk of early stopping (I'm not certain this is guaranteed).

**Against Early Stopping**

1. In some cases, results may not be reproducible.
2. In rare cases, the performance of the last model/s may be understated due to early stopping, which could result in user's being provided a less performant model through `automl.model`.

**Things to Note** 

- This early stopping will only ever affect the very last model (or group of models if using BlendSearch). I don't know how much time this may add onto the AutoML runtime, but I have to assume not very much.
- I've had to search for a specific configuration which triggers this issue - the chances of it occurring otherwise are low, albeit still possible.

Thanks for any help - would appreciate your opinions on the above (I know it's a lot). I'd still very much argue in favour of removing this early stopping behaviour, but can accept there might be a few good arguments for retaining it.

Reply:
Each time `_callback` is called, `EarlyStopException` will be called only when deadline is reached, that's why you see more calls on `_callback`. It's not necessary 4 times. If you change your parameters, you'll see different results.

![image](https://github.com/user-attachments/assets/0386024b-068c-49a8-bac1-2b8a2924eda1)
 
Deadline changes because `start_time` is updated at each `fit`.
![image](https://github.com/user-attachments/assets/b05aa7f4-87ca-4e96-8335-46a996f3bd84)

I think we should keep the time budget limitation here, otherwise, it could happen that the run exceeds the time budget too far. It'll happen a lot when the dataset is huge. In most cases, the last configuration wouldn't be picked up as the best parameters. Even it's picked, the retraining on it will get a better performance instead of worse one.